### PR TITLE
Refactoring the Windows specific code in NegotiateStream

### PR DESCRIPTION
**Pull request by shrutigarg**
summary 

Renaming NegoState.Windows.cs to InternalNegoState.cs as it had much common code applicable for Unix. (commit1)
Separating out the common code and windows specific code in InternalNegoState and NegoState.Windows.cs respectiviely.
Adding NegotiateInfoClass.cs to be used by Unix impl.

Tested locally both AuthenticateAsServer/Client on Windows.

/cc: @SidharthNabar @stephentoub  @davidsh @CIPop @vijaykota

### CODE REVIEW

**Comment by vijaykota** [comment]
cc: @joshfree to help track #2483

**Comment by stephentoub** [comment]
cc: @bartonjs

File: src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Net
+{
+    // This class is used to determine if NTLM or
+    // Kerberos are used in the context of a Negotiate handshake
+    internal class NegotiationInfoClass
```
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs]: Why is this in the Interop/.../Native directory?

File: src/System.Net.Security/src/System/Net/NTAuthentication.cs [hunk 2]
```diff
@@ -23,12 +23,12 @@ internal class NTAuthentication
...
         private int _tokenSize;
-        private Interop.SspiCli.ContextFlags _requestedContextFlags;
-        private Interop.SspiCli.ContextFlags _contextFlags;
+        private ContextFlagsPal _requestedContextFlags;
+        private ContextFlagsPal _contextFlags;
 
         private bool _isCompleted;
         private string _protocolName;
-        private SecSizes _sizes;
+        private object _sizes;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/NTAuthentication.cs]: Casting from object is always dangerous.  Is this really necessary?

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+            _readSequenceNumber = 0;
+
+            ContextFlagsPal flags = ContextFlagsPal.Connection;
+
+            // A workaround for the client when talking to Win9x on the server side.
+            if (protectionLevel == ProtectionLevel.None && !isServer)
+            {
+                package = NegotiationInfoClass.NTLM;
+            }
+
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Nit: Please delete the gap between the `if` and `else if` here.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+            else if (protectionLevel == ProtectionLevel.EncryptAndSign)
+            {
+                flags |= ContextFlagsPal.Confidentiality;
+            }
+            else if (protectionLevel == ProtectionLevel.Sign)
+            {
+                // Assuming user expects NT4 SP4 and above.
+                flags |= (ContextFlagsPal.ReplayDetect | ContextFlagsPal.SequenceDetect | ContextFlagsPal.InitIntegrity);
+            }
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: `if`/`else if`/`else if`/... waterfalls that don't end in an `else` always make me nervous.  Right now `ProtectionLevel.None` with `isServer` set doesn't do anything.  Are there other cases that don't do anything?  If not, might it be reasonable to add

```C#
else
{
    Debug.Assert(protectionLevel == ProtectionLevel.None);
}
```

So if a future protection mode comes online (or somehow the value 93 accidentally gets passed in) we can catch it.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+
+                if (_extendedProtectionPolicy.PolicyEnforcement != PolicyEnforcement.Never &&
+                    _extendedProtectionPolicy.ProtectionScenario == ProtectionScenario.TrustedProxy)
+                {
+                    flags |= ContextFlagsPal.ProxyBindings;
+                }
+            }
+            else
+            {
+                // Server side should not request any of these flags.
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: That comment sounds more like it would be on an assert in `if (isServer)` than parsing values in the else.

Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: (I think that's me saying "maybe we should add some asserts in the if" :smile:)

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+                    flags |= ContextFlagsPal.InitIdentify;
+                }
+
+                if (impersonationLevel == TokenImpersonationLevel.Delegation)
+                {
+                    flags |= ContextFlagsPal.Delegate;
+                }
+            }
+
+            _canRetryAuthentication = false;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: I know this was just moved... but as a reader of this code I have no idea why we just set canRetry to false.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+            {
+                throw new AuthenticationException(SR.net_auth_SSPI, e);
+            }
+        }
+
+        private Exception SetException(Exception e)
+        {
+            if (_exception == null || !(_exception is ObjectDisposedException))
+            {
+                _exception = e;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: @stephentoub I haven't traced the usage yet, but does this seem like it should be using `ExceptionDispatchInfo` instead?

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+            {
+                return _context != null && _context.IsServer;
+            }
+        }
+
+        internal bool CanGetSecureStream
+        {
+            get
+            {
+                return (_context.IsConfidentialityFlag || _context.IsIntegrityFlag);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: `IsAuthenticated` and `IsServer` both check `_context` for null.  Many other properties use `IsAuthenticated`, inheriting the nullcheck.  This one just dereferences it.

If `_context` being null is a legitimate state to be in (which it seems like it is, given that an instance method builds it) this should probably have a check, too.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 9]
```diff
@@ -0,0 +1,822 @@
...
+                return PrivateImpersonationLevel;
+            }
+        }
+
+        private TokenImpersonationLevel PrivateImpersonationLevel
+        {
+            get
+            {
+                // We should suppress the delegate flag in NTLM case.
+                return (_context.IsDelegationFlag && _context.ProtocolName != NegotiationInfoClass.NTLM) ? TokenImpersonationLevel.Delegation
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: A) Check _context for null.
B) Can this be rewritten to not be a ternary expression within a ternary expression?  It's... a bit hard to parse.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: https://github.com/dotnet/corefx/pull/5541#discussion_r50399913

:wink:

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 10]
```diff
@@ -0,0 +1,822 @@
...
+            }
+
+            if (authSucessCheck && !IsAuthenticated)
+            {
+                throw new InvalidOperationException(SR.net_auth_noauth);
+            }
+        }
+
+        //
+        // This is to not depend on GC&SafeHandle class if the context is not needed anymore.
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: I'm not sure I understand the comment here.  Aside from the overall not quite getting it, I don't know if that ampersand is a typo (GCSafeHandle) or is being used as "and", but without spacing around it.

And, no matter what, I don't see any SafeHandles here.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 11]
```diff
@@ -0,0 +1,822 @@
...
+        //
+        internal void Close()
+        {
+            // Mark this instance as disposed.
+            _exception = new ObjectDisposedException("NegotiateStream");
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: It seems weird that Close is going to allocate an Exception that may never get thrown.  But I'll let @stephentoub fight that battle :).

**Comment by joshfree** [comment]
cc @ericeil

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 12]
```diff
@@ -0,0 +1,822 @@
...
+
+            if (HandshakeComplete)
+            {
+                if (_context.IsServer && !CheckSpn())
+                {
+                    exception = new AuthenticationException(SR.net_auth_bad_client_creds_or_target_mismatch);
+                    int statusCode = ERROR_TRUST_FAILURE;
+                    message = new byte[8];  //sizeof(long)
+
+                    for (int i = message.Length - 1; i >= 0; --i)
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: I see three copies of this currently on my screen.  Perhaps factor it out to a method?

Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Actually, all three are doing the same runtime work to send ERROR_TRUST_FAILURE.  If that's at all common, maybe it should be a private static readonly field.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 13]
```diff
@@ -0,0 +1,822 @@
...
+                    throw;
+                }
+
+                lazyResult.InvokeCallback(e);
+            }
+        }
+
+        internal static bool IsError(SecurityStatusPal status)
+        {
+            return ((int)status >= (int)SecurityStatusPal.OutOfMemory);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Why the casts?

Is that enum guaranteed well-ordered?

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs [hunk 14]
```diff
@@ -0,0 +1,822 @@
...
+        {
+            SecurityStatusPal statusCode;
+            byte[] message = _context.GetOutgoingBlob(incomingBlob, false, out statusCode);
+
+            if (IsError(statusCode))
+            {
+                e = CreateExceptionFromError(statusCode);
+                uint error = (uint)e.HResult;
+
+                message = new byte[8];  //sizeof(long)
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Another copy of Uint64BigEndian

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 15]
```diff
@@ -13,869 +13,614 @@
...
-        private NTAuthentication _context;
+            protocol = _context.ProtocolName;
 
-        private int _nestedAuth;
+            if (_context.IsServer)
+            {
+                SecurityContextTokenHandle token = null;
+                try
+                {
+                    //token = _context.GetContextToken();
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: This method seems significantly different after the move.  Before it called GetContextToken then returned a new WindowsIdentity.  Now it does a lot more work, and the call to GetContextToken is commented out.

Maybe this is just inlining a method that got removed in the refactor? If so, this commented out line of code should at least be deleted.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 16]
```diff
@@ -13,869 +13,614 @@
...
 
-        internal static string DefaultPackage
+        internal static object QueryContextSizes(SafeDeleteContext securityContext)
         {
-            get
-            {
-                return NegotiationInfoClass.Negotiate;
-            }
+            return SSPIWrapper.QueryContextAttributes(
+              GlobalSSPI.SSPIAuth,
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: Nit: Seems like indentation is off here (parameters look inset by two spaces instead of four)

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 17]
```diff
@@ -13,869 +13,614 @@
...
+              securityContext,
+              Interop.SspiCli.ContextAttribute.Sizes
+              ) as SecSizes;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: If the return type has to stay object (which I hope not) there should be a comment as to why this is doing an `as`.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 18]
```diff
@@ -13,869 +13,614 @@
...
+            SecSizes sizes = secSizes as SecSizes;
 
-            if (isServer)
+            try
             {
-                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported)
-                {
-                    flags |= Interop.SspiCli.ContextFlags.AllowMissingBindings;
-                }
+                int maxCount = checked(Int32.MaxValue - 4 - sizes.BlockSize - sizes.SecurityTrailer);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: This will emit a NullReferenceException if the secSizes value had the wrong type.  Making me continue to question why it became `object` in the signature.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 19]
```diff
@@ -13,869 +13,614 @@
...
-            {
-                _context.CloseContext();
-            }
+            // Make a copy of user data for in-place encryption.
+            Buffer.BlockCopy(buffer, offset, output, 4 + sizes.SecurityTrailer, count);
 
-            return _exception;
-        }
+            // Prepare buffers TOKEN(signature), DATA and Padding.
+            var securityBuffer = new SecurityBuffer[3];
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: Rather than potentially having errors in the indices (and length), consider:

```C#
SecurityBuffer[] securityBuffer =
{
    new SecurityBuffer(output, 4, sizes.SecurityTrailer, SecurityBufferType.Token),
    new SecurityBuffer(output, 4 + sizes.SecurityTrailer, count, SecurityBufferType.Data),
    new SecurityBuffer(output, 4 + sizes.SecurityTrailer + count, sizes.BlockSize, SecurityBufferType.Padding),
};
```

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 20]
```diff
@@ -13,869 +13,614 @@
...
+            securityBuffer[0] = new SecurityBuffer(output, 4, sizes.SecurityTrailer, SecurityBufferType.Token);
+            securityBuffer[1] = new SecurityBuffer(output, 4 + sizes.SecurityTrailer, count, SecurityBufferType.Data);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: There are an awful lot of `4`s here.  Can we name that value?

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 19 above)
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: But since they get referenced by index everywhere else, maybe this doesn't matter.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 21]
```diff
@@ -13,869 +13,614 @@
...
+            out int newOffset,
+            uint sequenceNumber)
         {
-            CheckThrow(true);
-
-            IIdentity result = null;
-            string name = _context.IsServer ? _context.AssociatedName : _context.Spn;
-            string protocol = "NTLM";
+            var securityBuffer = new SecurityBuffer[2];
+            securityBuffer[0] = new SecurityBuffer(buffer, offset, 16, SecurityBufferType.Token);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: Okay, the 4s I thought I understood.  16 definitely needs a const.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 22]
```diff
@@ -13,869 +13,614 @@
...
 
-        internal void CheckThrow(bool authSucessCheck)
+        private static ContextFlagsPal GetContextFlagsPalFromInterop(Interop.SspiCli.ContextFlags win32Flags)
         {
-            if (_exception != null)
+            ContextFlagsPal flags = ContextFlagsPal.Zero;
+            if ((win32Flags & Interop.SspiCli.ContextFlags.AcceptExtendedError) != 0)
             {
-                throw _exception;
+                flags |= ContextFlagsPal.AcceptExtendedError;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: Consider a structure that you can just loop over, rather than repeating this many, many times.
```C#
private static readonly KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal>[] s_contextFlagMappings =
{
    new KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal>(Interop.SspiCli.ContextFlags.AcceptExtendedError, ContextFlagsPal.AcceptExtendedError),
    ...
};

...

ContextFlagsPal flags = ContextFlagsPal.Zero;

foreach (KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal> pair in s_contextFlagMappings)
{
    if ((win32Flags & pair.Key) != 0)
    {
        flags |= pair.Value;
    }
}
```

Perhaps with a using-alias to make the type identifier a wee-bit smaller.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs [hunk 23]
```diff
@@ -13,869 +13,614 @@
...
-        //
-        // Server side starts here, but client also loops through this method.
-        //
-        private void StartReceiveBlob(LazyAsyncResult lazyResult)
+        private static Interop.SspiCli.ContextFlags GetInteropFromContextFlagsPal(ContextFlagsPal flags)
         {
-            byte[] message;
-            if (lazyResult == null)
+            Interop.SspiCli.ContextFlags win32Flags = Interop.SspiCli.ContextFlags.Zero;
+            if ((flags & ContextFlagsPal.AcceptExtendedError) != 0)
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: This could also benefit from being turned into a loop.

File: src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs (hunk 1 above)
Comment (vijaykota) [inline src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs]: For Windows, it is under Interop\Windows\SspiCli. Any recommendations for Unix? (it is needed for both SPNEGO and NTLM protocols)

File: src/System.Net.Security/src/System/Net/NTAuthentication.cs (hunk 2 above)
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/NTAuthentication.cs]: @shrutigarg, if ```NegoState.Encrypt``` is the only consumer, we may be able to remove this property and add the required code there.

File: src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs (hunk 1 above)
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs]: I guess my question is "does this really vary across OS, or should it be something shared across them?"

Comment (vijaykota) [inline src/Common/src/Interop/Unix/System.Net.Security.Native/NegotiationInfoClass.cs]: Sorry.. misunderstood the original question. We can change this to a partial class if that is preferable. The constants can be shared and the constructor containing Windows specific code can remain under SspiCli

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs (hunk 13 above)
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Yes.. it has explicit initializations. Was that the question? The code this replaces was casting from Interop.SecurityStatus to int and checking if MSB is set (ie. (int)status == 0x80000000)

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 15 above)
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: Yes.. the comment needs to be removed. Since token can be fetched only on Windows, the code had to be refactored. On Unix, this will throw PlatformNotSupported in v1 since server support won't be there

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs (hunk 3 above)
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Using issue #5617 to track this and several other comments related to code that has only been moved in this PR

File: src/System.Net.Security/src/System/Net/NTAuthentication.cs (hunk 2 above)
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/NTAuthentication.cs]: Forgot to explain why this was changed. @bartonjs, SecSizes is used only in the Windows code and the type is not defined for Unix. So instead of a dummy type, we used object. @shrutigarg will look into possible refactoring of this.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 15 above)
Comment (shrutigarg) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: removed commented line

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 16 above)
Comment (shrutigarg) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: fixed.

File: src/System.Net.Security/src/System/Net/NTAuthentication.cs (hunk 2 above)
Comment (shrutigarg) [inline src/System.Net.Security/src/System/Net/NTAuthentication.cs]: yes encrypt is the only place it is referred. Can remove this property and fetched directly in Encrypt.

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 17 above)
Comment (shrutigarg) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: removed this . and directly fetched SecSize in Encrypt method

**Comment by shrutigarg** [comment]
@dotnet-bot test this please

**Comment by CIPop** [comment]
@shrutigarg I'd like to postpone merging this PR given multiple test failures for System.Net.Security: #4467, #5283 #5284. 

This is to avoid regressions as the one seen in #5642 especially since a lot of common code is getting refactored.

**Comment by vijaykota** [comment]
> I'd like to postpone merging this PR 

@joshfree, please note that actual xplat implementation would then be blocked by this PR

**Comment by CIPop** [comment]
>  please note that actual xplat implementation would then be blocked by this PR

@shrutigarg @vijaykota What I meant is that only merging this to master should be blocked until I stabilize and enable the Windows tests. 
You can still follow the development pattern where a private branch is used for various PRs and reviews. When we have the tests working we can merge from that. Since this is a big item you may also want to request a branch within dotnet/corefx.

**Comment by vijaykota** [comment]
Understood, @CIPop  We are merely informing @joshfree since he is tracking the final merge to master for consumption by partners.

> given multiple test failures for System.Net.Security: #4467, #5283 #5284. 

I take it you are actively working on code changes for these issues. Do you have an ETA for your changes getting merged so we can calculate our ETA?

File: src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs (hunk 18 above)
Comment (shrutigarg) [inline src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs]: already fixed this.

**Comment by vijaykota** [comment]
@bartonjs, please take a look at the changes made by @shrutigarg and comment if anything not covered by #5617 needs changes.

File: src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs (hunk 13 above)
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs]: Okay, I guess that it does have a comment indicating that OutOfMemory and beyond are "Errors", but since there's no gap in that enumeration I'm concerned about future expansion needs.

Since this code is assuming that everything with a higher value than SecurityStatusPal.OutOfMemory is an error, there really should be a gap, so possibly assign it equal to 0x010000 (or whatever) to leave room for later.

The casts seem completely unnecessary to me... enums are comparable with `>=`, so this line should be able to be `return (status >= SecurityStatusCode.OutOfMemory);`.

**Comment by vijaykota** [comment]
@stephentoub , PTAL There is one comment pending in IsError which @shrutigarg  will fix

**Comment by stephentoub** [comment]
@CIPop, we would like to move the implementation of the unix side of NegotiateStream into a separate branch, and I've created dev/negotiatestream for that purpose.  Do you have any plans to make significant changes to the Windows implementation in the near future?  If so, it would seem pragmatic to merge into master whatever changes are necessary to refactor that code so that there won't be major conflicts between the work done in master and the work done in the separate branch.  Or, are the changes you're talking about only impacting the tests, such that this refactoring can be done in the separate branch without causing significant merge issues?

**Comment by CIPop** [comment]
> Do you have any plans to make significant changes to the Windows implementation in the near future? 

@stephentoub As far as I can tell right now, we have everything we need from Desktop and no major changes will be made to the production code.
The only thing that I'll try to get to this week is fixing the test code which, we already know, has some design bugs.

**Comment by stephentoub** [comment]
Thanks, @cipop.  Sounds then like there should be few if any concerns around merge conflicts (if anywhere they'd be in tests), and we can stabilize everything in the separate branch before merging it into master.  @vijaykota, sound good?  We can take very frequent merges from master to the dev branch to ensure it's up-to-date, and @cipop, if we do have any unexpected changes to NegotiateStream that need to get made in the interim, it'd be great if you could help ensure that such merges are done correctly.

**Comment by vijaykota** [comment]
>  @vijaykota, sound good? We can take very frequent merges from master to the dev branch to ensure it's up-to-date

Thanks, @stephentoub  and @CIPop This should be good enough for us to branch off the work (including this PR) to the dev branch.

**Comment by vijaykota** [comment]
Added a label to prevent accidental merge while we are moving this to dev/negotiatestream

**Comment by shrutigarg** [comment]
moved this to #5772 in dev/NegotiateStream branch .
